File: medieval.py
import sys
import queue as Q
sys.stdin = open('input.txt')


def prim(N):
    global adj
    pq = Q.PriorityQueue()
    MSTEdges = 0
    total_weight = 0
    visited = [False for i in range(N+1)]
    visited[1] = True
    for edge in adj[1]:
        pq.put((edge[1], 1, edge[0])) # weight, start node, end node
    while MSTEdges < N and not pq.empty():
        state = pq.get()
        weight = state[0]
        start_node = state[1]
        end_node = state[2]
        if visited[start_node] and not visited[end_node]:
            visited[end_node] = True
            MSTEdges += 1
            total_weight += weight
            for edge in adj[end_node]:
                pq.put((edge[1], end_node, edge[0]))  # weight, start node, end node
    #print(MSTEdges, total_weight)
    if MSTEdges == N-1:
        return total_weight
    else:
        return -1


if __name__ == '__main__':
    T = int(input())
    adj = {}
    for t in range(T):
        N, C = map(int, input().split())
        for i in range(N+1):
            adj[i] = []
        for i in range(C):
            u, v, w = map(int, input().split())
            adj[u].append((v, w))
            adj[v].append((u, w))
        #print(adj)
        ans = prim(N)
        if ans == -1:
            print('Requires more conduits')
        else:
            print(ans)
        adj.clear()

File: medieval.cpp
#include<iostream>
#include<cstdio>
#include<cstring>
#include<algorithm>
#include<queue>
using namespace std;

const int MAXN = 100, INF = 0x3F3F3F3F;
int T,N,C;
bool visited[MAXN];
vector< pair<int,int> >adj[MAXN];

int prim(){
    priority_queue< pair<int, pair<int,int> > >pq;    // (edge){weight,start node,end node}
    int MSTEdges = 0, totalWeight = 0;
    // note that in C++ the priority queue is by default a max heap,
    // so a good trick to make it a min heap is when inserting items
    // into the heap, we negate them, then when we retrieve items,
    // we negate it again
    for(int i = 0; i < adj[0].size();i++){// arbitarily choose first node to start prim's algorithm from, 0 is the first node;
        pq.push(make_pair(-adj[0][i].second, make_pair(0,adj[0][i].first)));// {weight, starting node, ending node}
    }
    memset(visited,false,sizeof visited);// set all nodes to not visited
    visited[0] = true;// set first node to visisted
    while(MSTEdges < N-1 && !pq.empty()){// Minimum spanning trees have N-1 edges
        int w = -pq.top().first;
        int sn = pq.top().second.first;
        int en = pq.top().second.second;
        pq.pop();
        if(!visited[en]){
            visited[en] = true;
            MSTEdges++;
            totalWeight+=w;
            for(int i = 0; i < adj[en].size();i++)
                pq.push(make_pair(-adj[en][i].second, make_pair(en,adj[en][i].first)));// {weight, starting node, ending node}
        }
    }
    return (MSTEdges == N-1) ? totalWeight:-1;// if we have less than N-1 edges then that means the graph is disconnected
}
int main(){
    freopen("input.txt","r",stdin);
    scanf("%d",&T);
    while(T--){
        scanf("%d%d",&N,&C);
        for(int c = 0,i,j,k;c < C;c++){
            scanf("%d%d%d",&i,&j,&k);
            i--;j--; // moving everything over to start from 0 through to N-1
            // construct the adjacency matrix
            adj[i].push_back(make_pair(j,k));
            adj[j].push_back(make_pair(i,k));
        }
        int ans = prim();
        if(ans != -1)
            printf("%d\n",ans);
        else
            printf("Requires more conduits\n");
        for(int i = 0; i < MAXN;i++)
            adj[i].clear();
    }
    return 0;
}
